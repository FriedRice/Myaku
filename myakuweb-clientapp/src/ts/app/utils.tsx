/**
 * Utility functions for the MyakuWeb project.
 * @module utils
 */

import {
    PrimativeType,
    isIndexable,
    isPrimativeType,
} from 'ts/types/types';

interface MonthMap {
    [key: number]: string;
}

const MONTH_SHORT_NAME_MAP: MonthMap = {
    0: 'Jan',
    1: 'Feb',
    2: 'Mar',
    3: 'Apr',
    4: 'May',
    5: 'Jun',
    6: 'Jul',
    7: 'Aug',
    8: 'Sep',
    9: 'Oct',
    10: 'Nov',
    11: 'Dec',
};

/**
 * Recursively transform all primative values of obj using transformFunc.
 *
 * Performs the transformation in-place in the given object.
 *
 * @param obj - Object to transform all values of.
 * @param transformFunc - Function to use to transform the primative values of
 * obj.
 * @param condFunc? - If given, only the key-value pairs of obj that return
 * true when given to this function will have their values transformed.
 */
export function recursivelyTransform(
    obj: unknown,
    transformFunc: (value: PrimativeType) => unknown,
    condFunc?: (key: string, value: PrimativeType) => boolean
): void {
    if (!isIndexable(obj)) {
        return;
    }

    for (const [key, value] of Object.entries(obj)) {
        if (isIndexable(value) && typeof value !== 'function') {
            recursivelyTransform(value, transformFunc, condFunc);
        } else if (
            isPrimativeType(value) && (!condFunc || condFunc(key, value))
        ) {
            obj[key] = transformFunc(value);
        }
    }
}

export function scrollToTop(): void {
    window.scrollTo(0, 0);
}

/**
 * Forces a redraw of the element by the browser.
 *
 * @param element - Element to force a redraw for.
 *
 * @returns The offsetHeight of the element, but this can be ignored. It's only
 * returned to prevent the access of offsetHeight property from being removed
 * by the compiler.
 */
export function reflow(element: HTMLElement): number {
    return element.offsetHeight;
}

/**
 * Get the number of full days between two Dates.
 */
export function getDaysBetween(d1: Date, d2: Date): number {
    var d1DayOnly = new Date(d1.getFullYear(), d1.getMonth(), d1.getDate());
    var d2DayOnly = new Date(d2.getFullYear(), d2.getMonth(), d2.getDate());

    var milliSecsBetween = Math.abs(d1DayOnly.getTime() - d2DayOnly.getTime());
    return milliSecsBetween / (1000 * 60 * 60 * 24);
}

/**
 * Get the ordinal suffix for a number (e.g. 1 -> st, 2 -> nd, ...).
 *
 * @param num - Number to get the ordinal suffix for.
 *
 * @returns The ordinal suffix for the number.
 */
function getOrdinalSuffix(num: number): string {
    var mod100 = num % 100;
    if (mod100 > 10 && mod100 < 20) {
        return 'th';
    }

    var mod10 = num % 10;
    if (mod10 === 1) {
        return 'st';
    } else if (mod10 === 2) {
        return 'nd';
    } else if (mod10 === 3) {
        return 'rd';
    } else {
        return 'th';
    }
}

/**
 * Convert the Date to a string with a nice human-readable date.
 *
 * @param date - Date to convert.
 *
 * @returns If the datetime is within a week of now, a string stating how many
 * days ago the datetime was (i.e. today, yesterday, 2 days ago, ...). If the
 * datetime is not within a week, a formatted date string in the form
 * "Jan 1st, 2019".
 */
export function humanizeDate(date: Date): string {
    var daysBetweenNow = getDaysBetween(new Date(), date);
    if (daysBetweenNow === 0) {
        return 'Today';
    } else if (daysBetweenNow === 1) {
        return 'Yesterday';
    } else if (daysBetweenNow <= 7) {
        return `${daysBetweenNow} days ago`;
    } else {
        var day = `${date.getDate()}${getOrdinalSuffix(date.getDate())}`;
        var month = MONTH_SHORT_NAME_MAP[date.getMonth()];
        var year = date.getFullYear();
        return `${month} ${day}, ${year}`;
    }
}
