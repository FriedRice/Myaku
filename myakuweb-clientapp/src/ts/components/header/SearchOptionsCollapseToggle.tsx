/**
 * SearchOptionsCollapseToggle component module. See
 * [[SearchOptionsCollapseToggle]].
 */

import React from 'react';

/** Props for the [[SearchOptionsCollapseToggle]] component. */
interface SearchOptionsCollapseToggleProps {
    /**
     * Whether the search options are currently collapsed or not.
     *
     * @remarks
     * The toggle will indicate that the search options will be shown by
     * toggling if this is true, and the toggle will indicate that the search
     * options will be hidden by toggling if this is false.
     */
    collapsed: boolean;

    /** Callback to call whether the toggle is clicked. */
    onToggle: () => void;
}
type Props = SearchOptionsCollapseToggleProps;

/**
 * Toggle button component for showing/hiding the search options in a form.
 *
 * @param props - See [[SearchOptionsCollapseProps]].
 */
const SearchOptionsCollapseToggle: React.FC<Props> = function(props) {
    var buttonText;
    if (props.collapsed) {
        buttonText = 'Show search options';
    } else {
        buttonText = 'Hide search options';
    }

    return (
        <button
            className='button-link search-options-toggle'
            type='button'
            onClick={props.onToggle}
        >
            {buttonText}
        </button>
    );
};

export default SearchOptionsCollapseToggle;
