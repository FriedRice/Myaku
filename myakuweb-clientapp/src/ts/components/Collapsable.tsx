/** @module Wrapper for a collapsable component */

import React from 'react';
import { reflow } from '../utils';

interface CollapsableProps {
    children: React.ReactNode;
    collapsed: boolean;
    onAnimationStart?: () => void;
    onAnimationEnd?: () => void;
}
type Props = CollapsableProps;

interface CollapsableState {
    collapsed: boolean;
    animationStarted: boolean;
}
type State = CollapsableState;

interface CollapsableStyle {
    height?: string;
}


class Collapsable extends React.Component<Props, State> {
    private _collapsableRef: React.RefObject<HTMLDivElement>;

    constructor(props: Props) {
        super(props);
        this.bindEventHandlers();
        this._collapsableRef = React.createRef();

        this.state = {
            collapsed: this.props.collapsed,
            animationStarted: false,
        };
    }

    componentDidUpdate(): void {
        var collapsable = this._collapsableRef.current;
        if (
            this.state.animationStarted
            || collapsable === null
            || this.props.collapsed === this.state.collapsed
        ) {
            return;
        }

        reflow(collapsable);
        this.setState({
            animationStarted: true,
        });
        if (this.props.onAnimationStart) {
            this.props.onAnimationStart();
        }
    }

    bindEventHandlers(): void {
        this.handleAnimationEnd = this.handleAnimationEnd.bind(this);
    }

    handleAnimationEnd(): void {
        if (this._collapsableRef.current !== null) {
            this._collapsableRef.current.removeEventListener(
                'transitionend', this.handleAnimationEnd
            );
        }
        this.setState({
            collapsed: this.props.collapsed,
            animationStarted: false,
        });
        if (this.props.onAnimationEnd) {
            this.props.onAnimationEnd();
        }
    }

    getNoAnimationClassList(): string[] {
        if (this.props.collapsed) {
            return ['collapse'];
        } else {
            return ['collapse', 'show'];
        }
    }

    getAnimationClassList(): string[] {
        if (this.props.collapsed && !this.state.animationStarted) {
            return ['collapse', 'show'];
        } else {
            return ['collapsing'];
        }
    }

    getAnimationInlineStyle(): CollapsableStyle {
        var collapsable = this._collapsableRef.current;
        if (collapsable === null) {
            return {};
        }

        if (!this.props.collapsed) {
            return {
                height: collapsable.scrollHeight + 'px',
            };
        } else if (!this.state.animationStarted) {
            return {
                height: collapsable.getBoundingClientRect()['height'] + 'px',
            };
        } else {
            return {};
        }
    }

    addAnimationEndListener(): void {
        var collapsable = this._collapsableRef.current;
        if (collapsable === null) {
            return;
        }

        collapsable.addEventListener(
            'transitionend', this.handleAnimationEnd
        );
    }

    render(): React.ReactElement {
        var classList: string[] = [];
        var style: CollapsableStyle = {};
        if (this.props.collapsed === this.state.collapsed) {
            classList = this.getNoAnimationClassList();
        } else {
            classList = this.getAnimationClassList();
            style = this.getAnimationInlineStyle();
            if (!this.state.animationStarted) {
                this.addAnimationEndListener();
            }
        }

        return (
            <div
                className={classList.join(' ')}
                style={style}
                ref={this._collapsableRef}
            >
                {this.props.children}
            </div>
        );
    }
}

export default Collapsable;
