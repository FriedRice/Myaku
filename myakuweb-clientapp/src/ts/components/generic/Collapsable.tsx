/**
 * Collapsable component module. See [[Collapsable]].
 */

import React from 'react';
import { reflow } from 'ts/app/utils';

/** Props for the [[Collasable]] component. */
interface CollapsableProps {
    /** Child nodes to make collapsable. */
    children: React.ReactNode;

    /**
     * If true, the component will be collapsed (hiding the children), and
     * if false the component will be uncollapsed (showing the children).
     */
    collapsed: boolean;

    /**
     * If true, changes between collapsed and uncollapsed will be animated, and
     * if false, those changes will be instant with no animation.
     */
    animate: boolean;

    /**
     * Callback function that will be called whenever a collapse or uncollapse
     * animation starts.
     */
    onAnimationStart?: () => void;

    /**
     * Callback function that will be called whenever a collapse or uncollapse
     * animation ends.
     */
    onAnimationEnd?: () => void;
}
type Props = CollapsableProps;

/** State for the [[Collapsable]] component. */
interface CollapsableState {
    /**
     * Whether the collapsable is currently collapsed (true) or uncollapsed
     * (false).
     *
     * @remarks
     * The collapsed prop value indicates the current desired collapsed state,
     * but the collapsed state value indicates the actual current collapsed
     * state.
     * If these two values differ, the component will work to change its
     * collapse/uncollapse state (possibly with an animation depending on the
     * animate prop value) so that the state value will then match the prop
     * value.
     */
    collapsed: boolean;

    /**
     * Whether a collapse/uncollapse animation has started (true) or not
     * (false).
     */
    animationStarted: boolean;
}
type State = CollapsableState;

/** CSS style modified in the process of collapse/uncollapse animation */
interface CollapsableStyle {
    /**
     * The height must be explicitly set inline to trigger the animation at
     * times.
     */
    height?: string;
}


/**
 * Wrapper component to make another component collapsable.
 *
 * @remarks
 * See [[CollapsableProps]] and [[CollapsableState]] for props and state
 * details.
 */
class Collapsable extends React.Component<Props, State> {
    /** Ref for the collapsable container element. */
    private _collapsableRef: React.RefObject<HTMLDivElement>;

    constructor(props: Props) {
        super(props);
        this.bindEventHandlers();
        this._collapsableRef = React.createRef();

        this.state = {
            collapsed: this.props.collapsed,
            animationStarted: false,
        };
    }

    /**
     * Start the collapse/uncollapse animation if necessary.
     *
     * @remarks
     * If the collapsed prop and state values do not currently match and no
     * animation has been started, starts the collapse/uncollapse animation so
     * that the collapsed state value will match the prop value after the
     * animation.
     *
     * If the animate prop is false, just immediately sets the collapsed state
     * value to match the prop value.
     */
    componentDidUpdate(): void {
        var collapsable = this._collapsableRef.current;
        if (
            this.state.animationStarted
            || collapsable === null
            || this.props.collapsed === this.state.collapsed
        ) {
            return;
        }

        if (this.props.animate) {
            reflow(collapsable);
            this.setState({
                animationStarted: true,
            });
            if (this.props.onAnimationStart) {
                this.props.onAnimationStart();
            }
        } else {
            this.setState({
                collapsed: this.props.collapsed,
            });
        }
    }

    /**
     * Bind "this" for the event handlers used by the component.
     */
    bindEventHandlers(): void {
        this.handleAnimationEnd = this.handleAnimationEnd.bind(this);
    }

    /**
     * Callback for updating component state once a collapse/uncollapse
     * animation ends.
     */
    handleAnimationEnd(): void {
        if (this._collapsableRef.current !== null) {
            this._collapsableRef.current.removeEventListener(
                'transitionend', this.handleAnimationEnd
            );
        }
        this.setState({
            collapsed: this.props.collapsed,
            animationStarted: false,
        });
        if (this.props.onAnimationEnd) {
            this.props.onAnimationEnd();
        }
    }

    /**
     * Get the no-animation class list for the collapsable element based on the
     * current props.
     *
     * @returns The list of classes.
     */
    getNoAnimationClassList(): string[] {
        if (this.props.collapsed) {
            return ['collapse'];
        } else {
            return ['collapse', 'show'];
        }
    }

    /**
     * Get the animation class list for the collapsable element based on the
     * current props and state.
     *
     * @returns The list of classes.
     */
    getAnimationClassList(): string[] {
        if (this.props.collapsed && !this.state.animationStarted) {
            return ['collapse', 'show'];
        } else {
            return ['collapsing'];
        }
    }

    /**
     * Get the inline CSS style for the collapsable element based on the
     * current props and state.
     *
     * @returns The inline CSS styles.
     */
    getAnimationInlineStyle(): CollapsableStyle {
        var collapsable = this._collapsableRef.current;
        if (collapsable === null) {
            return {};
        }

        if (!this.props.collapsed) {
            return {
                height: collapsable.scrollHeight + 'px',
            };
        } else if (!this.state.animationStarted) {
            return {
                height: collapsable.getBoundingClientRect()['height'] + 'px',
            };
        } else {
            return {};
        }
    }

    /**
     * Set a callback to be called on a collapse/uncollapse animation end to
     * update component state.
     */
    addAnimationEndListener(): void {
        var collapsable = this._collapsableRef.current;
        if (collapsable === null) {
            return;
        }

        collapsable.addEventListener(
            'transitionend', this.handleAnimationEnd
        );
    }

    render(): React.ReactElement {
        var classList: string[] = [];
        var style: CollapsableStyle = {};
        if (
            this.props.collapsed === this.state.collapsed
            || !this.props.animate
        ) {
            classList = this.getNoAnimationClassList();
        } else {
            classList = this.getAnimationClassList();
            style = this.getAnimationInlineStyle();
            if (!this.state.animationStarted) {
                this.addAnimationEndListener();
            }
        }

        return (
            <div
                className={classList.join(' ')}
                style={style}
                ref={this._collapsableRef}
            >
                {this.props.children}
            </div>
        );
    }
}

export default Collapsable;
