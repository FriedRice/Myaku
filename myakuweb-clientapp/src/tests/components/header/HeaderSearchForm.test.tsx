/**
 * Tests for the [[HeaderSearchForm]] component.
 */

import Collapsable from 'ts/components/generic/Collapsable';
import HeaderSearchForm from 'ts/components/header/HeaderSearchForm';
import React from 'react';
import SearchBarInput from 'ts/components/header/SearchBarInput';
import { SearchOptions } from 'ts/types/types';
import SearchOptionsCollapseToggle from
    'ts/components/header/SearchOptionsCollapseToggle';
import SearchOptionsInput from 'ts/components/header/SearchOptionsInput';
import { act } from 'react-dom/test-utils';
import { blurActiveElement } from 'ts/app/utils';
import { createMemoryHistory } from 'history';
import { expectComponent } from 'tests/testUtils';

import {
    ShallowWrapper,
    shallow,
} from 'enzyme';
import {
    loadUserSearchOptions,
    setUserSearchOption,
} from 'ts/app/search';

const mockLoadUserSearchOptions = loadUserSearchOptions as jest.Mock;
const mockSetUserSearchOption = setUserSearchOption as jest.Mock;
const mockBlurActiveElement = blurActiveElement as jest.Mock;
const mockPromise = {
    then: (): void => {},
    catch: (): void => {},
};

/**
 * Mock only the async functions from the search module to avoid getting errors
 * when these functions get called in HeaderSearchForm during synchronous
 * tests.
 */
jest.mock('ts/app/search', function() {
    const original = jest.requireActual('ts/app/search');

    return {
        __esModule: true,
        ...original,
        loadUserSearchOptions: jest.fn(function() {
            return mockPromise;
        }),
        setUserSearchOption: jest.fn(function() {
            return mockPromise;
        }),
    };
});

jest.mock('ts/app/utils', function() {
    const original = jest.requireActual('ts/app/utils');

    return {
        __esModule: true,
        ...original,
        blurActiveElement: jest.fn(() => {}),
    };
});

var history = createMemoryHistory();
var wrapper = shallow(<div />);
beforeEach(function() {
    mockLoadUserSearchOptions.mockClear();
    mockSetUserSearchOption.mockClear();
    mockBlurActiveElement.mockClear();

    history = createMemoryHistory();
    wrapper = shallow(
        <HeaderSearchForm
            loadingSearch={false}
            location={history.location}
            history={history}
        />
    );
});


function expectSearchBarLoading(
    wrapper: ShallowWrapper, loading: boolean
): void {
    expectComponent(wrapper, SearchBarInput, {
        searchQuery: expect.any(String),
        loading: loading,
        maxQueryLength: expect.any(Number),
        errorValueSubmitted: false,
        onChange: expect.any(Function),
    });
}

function expectSearchBarQuery(
    wrapper: ShallowWrapper, query: string
): void {
    expectComponent(wrapper, SearchBarInput, {
        searchQuery: query,
        loading: false,
        maxQueryLength: expect.any(Number),
        errorValueSubmitted: false,
        onChange: expect.any(Function),
    });
}

function expectSearchBarError(
    wrapper: ShallowWrapper, errorValueSubmitted: boolean
): void {
    expectComponent(wrapper, SearchBarInput, {
        searchQuery: expect.any(String),
        loading: false,
        maxQueryLength: expect.any(Number),
        errorValueSubmitted: errorValueSubmitted,
        onChange: expect.any(Function),
    });
}

function expectSearchOptionsCollapsed(
    wrapper: ShallowWrapper, collapsed: boolean,
    animate: boolean = expect.any(Boolean)
): void {
    expectComponent(wrapper, Collapsable, {
        collapsed: collapsed,
        animate: animate,
        onAnimationEnd: expect.any(Function),
        children: expect.anything(),
    });
}

function expectSearchOptions(
    wrapper: ShallowWrapper, options: SearchOptions
): void {
    expectComponent(wrapper, SearchOptionsInput, {
        searchOptions: options,
        onChange: expect.any(Function),
    });
}

function expectStartHistory(): void {
    expect(history).toHaveLength(1);
    expect(history.location.pathname).toBe('/');
    expect(history.location.search).toBe('');
}

function expectSearchHistoryLocation(
    query: string, historyLength: number
): void {
    expect(history).toHaveLength(historyLength);
    expect(history.location.pathname).toBe('/search/');
    expect(history.location.search).toBe(`?q=${query}&p=1`);
}

function toggleSearchOptionsCollapse(wrapper: ShallowWrapper): void {
    const collapseToggle = wrapper.find(SearchOptionsCollapseToggle);
    expect(collapseToggle).toHaveLength(1);
    act(() => collapseToggle.props().onToggle());
}

function uncollapseSearchOptions(wrapper: ShallowWrapper): void {
    expectSearchOptionsCollapsed(wrapper, true);
    toggleSearchOptionsCollapse(wrapper);
    expectSearchOptionsCollapsed(wrapper, false);
}

function mockCollapseOnAnimationEndEvent(wrapper: ShallowWrapper): void {
    const collapsable = wrapper.find(Collapsable);
    expect(collapsable).toHaveLength(1);

    const onAnimationEnd = collapsable.props().onAnimationEnd;
    if (onAnimationEnd === undefined) {
        throw new Error(
            'Search options collapsable has no onAnimationEnd callback set'
        );
    }
    act(() => onAnimationEnd());
}

function inputSearchQuery(wrapper: ShallowWrapper, query: string): void {
    const searchBarInput = wrapper.find(SearchBarInput);
    expect(searchBarInput).toHaveLength(1);
    act(() => searchBarInput.props().onChange(query));
}

function inputSearchOption<K extends keyof SearchOptions>(
    wrapper: ShallowWrapper, option: K, value: SearchOptions[K]
): void {
    const searchOptionsInput = wrapper.find(SearchOptionsInput);
    expect(SearchOptionsInput).toHaveLength(1);
    act(() => searchOptionsInput.props().onChange(option, value));
}

async function inputSearchOptionAsync<K extends keyof SearchOptions>(
    wrapper: ShallowWrapper, option: K, value: SearchOptions[K]
): Promise<void> {
    const searchOptionsInput = wrapper.find(SearchOptionsInput);
    expect(SearchOptionsInput).toHaveLength(1);
    act(() => searchOptionsInput.props().onChange(option, value));
}

function simulateMockSubmitEvent(wrapper: ShallowWrapper): jest.Mock {
    const mockPreventDefault = jest.fn();
    wrapper.simulate('submit', {preventDefault: mockPreventDefault});

    return mockPreventDefault;
}

async function getShallowWrapperAsync(): Promise<ShallowWrapper> {
    return shallow(
        <HeaderSearchForm
            loadingSearch={false}
            location={history.location}
            history={history}
        />
    );
}


describe('<HeaderSearchForm /> mount/unmount/page change', function() {
    it('renders correctly', function() {
        expect(wrapper).toMatchSnapshot();
    });

    it('sets rendered components as loading when loading', function() {
        expectSearchBarLoading(wrapper, false);

        wrapper.setProps({loadingSearch: true});
        expectSearchBarLoading(wrapper, true);
    });

    it('sets search query from location on mount', function() {
        act(() => history.push('/'));
        wrapper = shallow(
            <HeaderSearchForm
                loadingSearch={false}
                location={history.location}
                history={history}
            />
        );
        expectSearchBarQuery(wrapper, '');

        act(() => history.push('/search/?q=力士'));
        wrapper = shallow(
            <HeaderSearchForm
                loadingSearch={false}
                location={history.location}
                history={history}
            />
        );
        expectSearchBarQuery(wrapper, '力士');

        act(() => history.push('/search/?q=OB&p=1'));
        wrapper = shallow(
            <HeaderSearchForm
                loadingSearch={false}
                location={history.location}
                history={history}
            />
        );
        expectSearchBarQuery(wrapper, 'OB');
    });

    it('sets search query from location on history change', function() {
        act(() => history.push('/search/?q=力士'));
        expectSearchBarQuery(wrapper, '力士');

        act(() => history.push('/search/?q=OB&p=1'));
        expectSearchBarQuery(wrapper, 'OB');

        act(() => history.push('/'));
        expectSearchBarQuery(wrapper, '');
    });

    it('unlistens to history changes on unmount', function() {
        wrapper.unmount();
        expect(() => history.push('/search/?q=OB')).not.toThrow();
    });

    it('collapses search options by default on mount', function() {
        expectSearchOptionsCollapsed(wrapper, true, false);
    });

    it('resets search options to collapsed on history change', function() {
        uncollapseSearchOptions(wrapper);
        history.push('/search/?q=OB');
        expectSearchOptionsCollapsed(wrapper, true, false);

        uncollapseSearchOptions(wrapper);
        history.goBack();
        expectSearchOptionsCollapsed(wrapper, true, false);

        uncollapseSearchOptions(wrapper);
        history.goForward();
        expectSearchOptionsCollapsed(wrapper, true, false);

        uncollapseSearchOptions(wrapper);
        history.replace('/');
        expectSearchOptionsCollapsed(wrapper, true, false);
    });

    it('loads session search options on mount', async function() {
        mockLoadUserSearchOptions.mockResolvedValue({kanaConvertType: 'hira'});
        const wrapperHira = await getShallowWrapperAsync();
        expectSearchOptions(wrapperHira, {kanaConvertType: 'hira'});

        mockLoadUserSearchOptions.mockResolvedValue({kanaConvertType: 'kata'});
        const wrapperKata = await getShallowWrapperAsync();
        expectSearchOptions(wrapperKata, {kanaConvertType: 'kata'});

        mockLoadUserSearchOptions.mockResolvedValue({kanaConvertType: 'none'});
        const wrapperNone = await getShallowWrapperAsync();
        expectSearchOptions(wrapperNone, {kanaConvertType: 'none'});
    });

    it('logs warning on mount if no IndexedDB available', async function() {
        const mockConsoleWarning = jest.spyOn(console, 'warn');
        mockConsoleWarning.mockImplementation(() => {});
        mockLoadUserSearchOptions.mockRejectedValueOnce(new Error('error'));

        await getShallowWrapperAsync();
        expect(mockConsoleWarning).toBeCalledTimes(1);
        expect(mockConsoleWarning).lastCalledWith(
            expect.stringContaining('Unable to use IndexedDB')
        );
    });
});

describe('<HeaderSearchForm /> form submit', function() {
    var form = shallow(<div />);
    beforeEach(function() {
        form = wrapper.find('form');
    });

    it('pushes search URL to history on valid form submit', function() {
        inputSearchQuery(wrapper, 'ゆっくり');
        simulateMockSubmitEvent(form);
        expectSearchHistoryLocation('ゆっくり', 2);

        inputSearchQuery(wrapper, 'ゆっくり');
        simulateMockSubmitEvent(form);
        expectSearchHistoryLocation('ゆっくり', 3);

        inputSearchQuery(wrapper, '力士');
        simulateMockSubmitEvent(form);
        expectSearchHistoryLocation('力士', 4);
    });

    it('stops query too long form submit', function() {
        var longStrList = [];
        for (let i = 0; i < 100; ++i) {
            longStrList.push('ゆっくり');
        }
        expectSearchBarError(wrapper, false);
        expectStartHistory();

        inputSearchQuery(wrapper, longStrList.join(''));
        simulateMockSubmitEvent(form);
        expectStartHistory();
        expectSearchBarError(wrapper, true);
    });

    it('stops no query form submit', function() {
        expectSearchBarError(wrapper, false);
        expectStartHistory();

        inputSearchQuery(wrapper, '');
        simulateMockSubmitEvent(form);
        expectStartHistory();
        expectSearchBarError(wrapper, true);
    });

    it('applies hiragana conversion on from submit', function() {
        inputSearchQuery(wrapper, 'yukkuri');
        inputSearchOption(wrapper, 'kanaConvertType', 'hira');
        simulateMockSubmitEvent(form);
        expectSearchHistoryLocation('ゆっくり', 2);
        expectSearchBarQuery(wrapper, 'ゆっくり');
    });

    it('applies katakana conversion on from submit', function() {
        inputSearchQuery(wrapper, 'yukkuri');
        inputSearchOption(wrapper, 'kanaConvertType', 'kata');
        simulateMockSubmitEvent(form);
        expectSearchHistoryLocation('ユックリ', 2);
        expectSearchBarQuery(wrapper, 'ユックリ');
    });

    it('applies none conversion on from submit', function() {
        inputSearchQuery(wrapper, 'yukkuri');
        inputSearchOption(wrapper, 'kanaConvertType', 'none');
        simulateMockSubmitEvent(form);
        expectSearchHistoryLocation('yukkuri', 2);
        expectSearchBarQuery(wrapper, 'yukkuri');
    });

    it('applies no query conversion if query is not romaji', function() {
        inputSearchQuery(wrapper, 'ユックリ');
        inputSearchOption(wrapper, 'kanaConvertType', 'hira');
        simulateMockSubmitEvent(form);
        expectSearchHistoryLocation('ユックリ', 2);
        expectSearchBarQuery(wrapper, 'ユックリ');

        inputSearchQuery(wrapper, 'ゆっくり');
        inputSearchOption(wrapper, 'kanaConvertType', 'kata');
        simulateMockSubmitEvent(form);
        expectSearchHistoryLocation('ゆっくり', 3);
        expectSearchBarQuery(wrapper, 'ゆっくり');
    });

    it('blurs focus on form submit', function() {
        inputSearchQuery(wrapper, 'ゆっくり');

        mockBlurActiveElement.mockClear();
        simulateMockSubmitEvent(form);
        expect(mockBlurActiveElement).toBeCalledTimes(1);
    });

    it('clears submit error on history change', function() {
        inputSearchQuery(wrapper, '');

        simulateMockSubmitEvent(form);
        expectSearchBarError(wrapper, true);
        act(() => history.push('/'));
        expectSearchBarError(wrapper, false);

        simulateMockSubmitEvent(form);
        expectSearchBarError(wrapper, true);
        act(() => history.goBack());
        expectSearchBarError(wrapper, false);

        simulateMockSubmitEvent(form);
        expectSearchBarError(wrapper, true);
        act(() => history.goForward());
        expectSearchBarError(wrapper, false);

        simulateMockSubmitEvent(form);
        expectSearchBarError(wrapper, true);
        act(() => history.replace('/search/?q=OB'));
        expectSearchBarError(wrapper, false);
    });

    it('clears submit error on inputted search query change', function() {
        inputSearchQuery(wrapper, '');
        simulateMockSubmitEvent(form);
        expectSearchBarError(wrapper, true);

        inputSearchQuery(wrapper, 'test');
        expectSearchBarError(wrapper, false);
    });

    it('prevents default of the form submit event', function() {
        inputSearchQuery(wrapper, 'ゆっくり');
        const mockPreventDefault = simulateMockSubmitEvent(form);
        expect(mockPreventDefault).toBeCalledTimes(1);
    });
});

describe('<HeaderSearchForm /> form event handlers', function() {
    it('handles inputted search query changes', function() {
        inputSearchQuery(wrapper, 'ゆっくり');
        expectSearchBarQuery(wrapper, 'ゆっくり');

        inputSearchQuery(wrapper, '');
        expectSearchBarQuery(wrapper, '');

        inputSearchQuery(wrapper, 'OB');
        expectSearchBarQuery(wrapper, 'OB');
    });

    it('handles inputted search option changes', function() {
        inputSearchOption(wrapper, 'kanaConvertType', 'hira');
        expectSearchOptions(wrapper, {kanaConvertType: 'hira'});

        inputSearchOption(wrapper, 'kanaConvertType', 'kata');
        expectSearchOptions(wrapper, {kanaConvertType: 'kata'});

        inputSearchOption(wrapper, 'kanaConvertType', 'none');
        expectSearchOptions(wrapper, {kanaConvertType: 'none'});
    });

    it('saves inputted search option changes to session', function() {
        inputSearchOption(wrapper, 'kanaConvertType', 'hira');
        expect(mockSetUserSearchOption).toBeCalledTimes(1);
        expect(mockSetUserSearchOption).lastCalledWith(
            'kanaConvertType', 'hira'
        );

        inputSearchOption(wrapper, 'kanaConvertType', 'kata');
        expect(mockSetUserSearchOption).toBeCalledTimes(2);
        expect(mockSetUserSearchOption).lastCalledWith(
            'kanaConvertType', 'kata'
        );

        inputSearchOption(wrapper, 'kanaConvertType', 'none');
        expect(mockSetUserSearchOption).toBeCalledTimes(3);
        expect(mockSetUserSearchOption).lastCalledWith(
            'kanaConvertType', 'none'
        );
    });

    it(
        'logs warning on search option change if no IndexedDB available',
        async function() {
            const mockConsoleWarning = jest.spyOn(console, 'warn');
            mockConsoleWarning.mockImplementation(() => {});
            mockSetUserSearchOption.mockRejectedValueOnce(new Error('error'));

            wrapper = await getShallowWrapperAsync();
            mockConsoleWarning.mockClear();
            await inputSearchOptionAsync(wrapper, 'kanaConvertType', 'kata');
            expect(mockConsoleWarning).toBeCalledTimes(1);
            expect(mockConsoleWarning).lastCalledWith(
                expect.stringContaining('Unable to use IndexedDB')
            );
        }
    );

    it('handles search options collapse toggle', function() {
        expectSearchOptionsCollapsed(wrapper, true, false);

        toggleSearchOptionsCollapse(wrapper);
        expectSearchOptionsCollapsed(wrapper, false, true);

        mockCollapseOnAnimationEndEvent(wrapper);
        toggleSearchOptionsCollapse(wrapper);
        expectSearchOptionsCollapsed(wrapper, true, true);

        mockCollapseOnAnimationEndEvent(wrapper);
        toggleSearchOptionsCollapse(wrapper);
        expectSearchOptionsCollapsed(wrapper, false, true);
    });

    it('ignores search options collapse toggle if animating', function() {
        expectSearchOptionsCollapsed(wrapper, true, false);

        toggleSearchOptionsCollapse(wrapper);
        expectSearchOptionsCollapsed(wrapper, false, true);

        toggleSearchOptionsCollapse(wrapper);
        expectSearchOptionsCollapsed(wrapper, false, true);

        mockCollapseOnAnimationEndEvent(wrapper);
        toggleSearchOptionsCollapse(wrapper);
        expectSearchOptionsCollapsed(wrapper, true, true);

        toggleSearchOptionsCollapse(wrapper);
        expectSearchOptionsCollapsed(wrapper, true, true);
    });
});
