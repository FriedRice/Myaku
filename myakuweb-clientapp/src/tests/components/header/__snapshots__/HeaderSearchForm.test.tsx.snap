// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HeaderSearchForm /> mount/unmount/page change renders correctly 1`] = `
<div
  className="search-container"
>
  <form
    className="search-form"
    onSubmit={[Function]}
  >
    <SearchBarInput
      errorValueSubmitted={false}
      loading={false}
      maxQueryLength={100}
      onChange={[Function]}
      searchQuery=""
    />
    <Collapsable
      animate={false}
      collapsed={true}
      onAnimationEnd={[Function]}
    >
      <SearchOptionsInput
        onChange={[Function]}
        searchOptions={
          Object {
            "kanaConvertType": "hira",
          }
        }
      />
    </Collapsable>
  </form>
  <div
    className="search-options-toggle-container"
  >
    <SearchOptionsCollapseToggle
      collapsed={true}
      onToggle={[Function]}
    />
  </div>
</div>
`;
